java/arcs/android/storage/service/BindingContextStats.kt

```diff
@@ -69,4 +75,37 @@ class BindingContextStatsImpl : BindingContextStatistics {
             }
         }
     }
+
+    override fun traceTransaction(tag: String?, block: () -> Unit) {
+        // TODO(ianchang): Inject Android system traces with [tag]
+        ++_transactions
+        block()
+        --_transactions
+    }
+
+    companion object {
+        private var _transactions = Transactions(Mutex())
+    }
+}
+
+data class Transactions(val mutex: Mutex? = null, var current: Int = 0, var peak: Int = 0) {
```

Comment:
It seems incorrect to provide a mutex to a data class. IMO either the class should not be considered a data class (and not be directly mutable) while maintaining its own mutex, or the class should be immutable and the mutex should exist within the BindingContext.  Also, if Transactions is immutable, you might be able to get away with using an atomic reference: `atomic<Transaction>(...)` from AtomicFu.

Reply:
Ah, the Transactions itself is mutable (overloading ++/--), all I wanted to do is binding a mutex to such a Transactions instance, that is a Transactions instance shares a new lock.

Based on your comment, I feel I'm supposed to use a normal class instead of a data class.
Will change it, thanks.

Reply:
It's okay to have a mutable data class, it just felt wrong to have a data class be aware of mutual exclusion contexts.

Reply:
Gotcha, data is just data, who wants to use it at what context should be irrelevant to it.

I just refactored the design (also changed to use atomic you mentioned)
Also wrote up some comments of using atomic with caution.

Please help review again, hopefully it's more kotlin-idiomatic now :-)